### ISSUE TITLE
Disruptor Pattern

### ISSUE THREAD

**Issue by npathai** [iluwatar/java-design-patterns#112]
Disruptor is an architectural pattern which is used to create applications that need very low latency.

Quoting [LMAX Exchange](https://github.com/LMAX-Exchange/disruptor/wiki/Introduction)
> Like a queue the purpose of the Disruptor is to move data (e.g. messages or events) between threads within the same process. However there are some key features that the Disruptor provides that distinguish it from a queue. They are:
 - Multicast events to consumers, with consumer dependency graph.
 - Pre-allocate memory for events.
 - Optionally lock-free.

It is based on Ring buffer than a Blocking queue, to increase the [Mechanical sympathy](http://mechanical-sympathy.blogspot.in/), a term coined by Martin Thompson. The crux of the term is to develop software that can take the most out of hardware taking into consideration the hardware architecture. Example using data structures that utilize hardware caches and are lock free.

Known Implementations:
 - https://github.com/LMAX-Exchange/disruptor
 - https://code.google.com/p/disruptor/

![Schematic of ring buffer](http://martinfowler.com/articles/images/lmax/disruptor.png)

**Comment by iluwatar** [comment]
Great find @npathai :+1:

**Comment by npathai** [comment]
@iluwatar Yup its an amazing pattern. I will keep adding newer ones. I like learning patterns and will contribute lot here. :)

**Comment by iluwatar** [comment]
Disruptor was recently mentioned in Spring blog here: https://spring.io/blog/2016/07/20/notes-on-reactive-programming-part-iii-a-simple-http-server-application

**Comment by iluwatar** [comment]
http://lmax-exchange.github.io/disruptor/

**Comment by iluwatar** [comment]
Sure, check here first https://github.com/iluwatar/java-design-patterns/wiki/01.-How-to-contribute

Study the existing design patterns. The new one is expected to follow the same technical structure.

**Comment by jomariabejo** [comment]
Thank you for talking me to the requirements for contributions. Before proposing the Disruptor pattern implementation, I'll thoroughly research the existing design patterns to ensure they correspond with the technical structure.
